### DISCUSSION TITLE
Remap keys on a keyboard

### DISCUSSION THREAD

**Discussion by GuloGuloDesu** [labwc/labwc#2028] [Q&A] [answered]
Hello,

Is it possible to remap keyboard keys in labwc?
I am having a problem with my numlock and working in VM's.  So I am trying to map my keypad keys with numlock on or off to always send they keypad numbers.
I have a small workaround in my terminal (kitty). I have remapped all of my keys with numlock off with the following config (kitty.conf), but I would like to do this for all applications.

> map KP_Insert send_text all 0
> map KP_END send_text all 1
> map KP_DOWN send_text all 2
> map KP_Next send_text all 3
> map KP_Left send_text all 4
> map KP_Begin send_text all 5
> map KP_Right send_text all 6
> map KP_Home send_text all 7
> map KP_Up send_text all 8
> map KP_Prior send_text all 9
> map KP_Delete send_text all .

Is there a way to do this in labwc?

I tried to look at the environment file in labwc and to see if I could map the keys using XKB_DEFAULT_OPTIONS, but I was unable to find a way to get it to work. I tried to configure XKB_DEFAULT_OPTIONS=numpad:pc wihtout success.

**Comment by tokyo4j** [comment]
Labwc doesn't directly support key remapping, but you can bind keys to virtual-keyboard clients like:
```xml
<keybind key="End"><action name="Execute" command="wtype 1"></keybind>
```
Or use keyboard remapping tools [listed here](https://wiki.archlinux.org/title/Input_remap_utilities).

**Comment by Consolatis** [comment, marked as answer]
I'd assume it should also be possible to modify some xkb layout and then use that one via `XKB_DEFAULT_LAYOUT`. `XKB_DEFAULT_OPTIONS=numpad:mac` might also work, see `man xkeyboard-config`.

**Comment by GuloGuloDesu** [comment]
Just tried with numpad:mac and it worked perfectly after a reconfigure.
I guess that I gave up too soon.

For anyone else that finds this. I edited my `~/.config/labwc/environment `to contain:

> XKB_DEFAULT_LAYOUT=us
> XKB_DEFAULT_OPTIONS=numpad:mac

This has fixed the issue where inside my proxmox VM's (connected through NoVNC using Firefox) the numlock would have to be off for the keypad to work in the VM. But the numlock would have to be on for my regular desktop to use keypad. By making these changes to the VM, I am able to use my keypad with the numlock on, and it works on my desktop as well.

**Reply by JSellers0** [reply]
The biggest issue I've found with this solution is the frequency at which that file is overwritten.
